### DISCUSSION TITLE
How can I build the project?

### DISCUSSION THREAD

**Discussion by Rohupt** [saschanaz/ime-rs#4] [General]
What should I do to link the Rust side to the C++ side of the project? Or more generally, what steps should I take to build a working product from the repository?

To be exact, I got linker errors during building:
![image](https://user-images.githubusercontent.com/34348907/158163306-89bf6655-6bd5-4e25-b53b-6864600f81e7.png)

**Comment by saschanaz** [comment]
This is most probably because of 32bit vs 64bit... Currently the project is configured to only support 64bit, so you need to choose x64/Debug.

A few more things:
* https://github.com/saschanaz/ime-rs/issues/1#issuecomment-918582423
* https://github.com/saschanaz/ime-rs/issues/1#issuecomment-919447763

**Comment by Rohupt** [comment]
Now I want to build this on Win32 configs. I will change the configurations myself, but could you give an outline of what should be done? I'm aware of the pre-build script, linkers and some of the tweaks, but in the end there are still errors.

**Reply by saschanaz** [reply]
The prebuild script also need to tell Rust to compile for 32bit, as it currently just do its default.

**Reply by Rohupt** [reply]
Let's say I want to build two versions of the dll (Win32 and x64) and register both of them together (because for some reasons the 32-bit dll won't work with x64 apps and vice versa), do I need to tweak the registry GUIDs, and if yes, which ones should I change and how to generate an acceptable GUID to change them to?

**Reply by saschanaz** [reply]
In that case I guess there should be some conditional compilation: https://doc.rust-lang.org/reference/conditional-compilation.html

Generating GUID is easy, DuckDuckGo search page even automatically generates one for you: https://duckduckgo.com/?t=ffab&q=guid&ia=answer

In this case I believe (but not sure, as I haven't tried) that these two should change.

https://github.com/saschanaz/ime-rs/blob/a64b17899943b7099629a61564d5b4bd9cda0bc6/rust/globals/src/lib.rs#L3-L15

**Reply by Rohupt** [reply]
Turns out it's quite complicated. According to [here](https://docs.microsoft.com/en-us/windows/win32/tsf/64-bit-platform-considerations) the CLSID keys should be identical. We need to call [ImmInstallIMEA](https://docs.microsoft.com/en-us/windows/win32/api/imm/nf-imm-imminstallimea) function to generate the registry key to share data for both the DLLs etc etc. I'll find out all about this and share it here later.

**Reply by Rohupt** [reply]
So... I don't actually have to change anything related to the GUIDs. Just create 2 versions of the DLL and install them into System32 and SysWOW64 (together with the dictionaries of course) and it's done.
